src/rust/lib/graph/src/component/node/port.rs

```diff
@@ -0,0 +1,319 @@
+    fn new_port_outwards(spec:SpecificationVar) -> AnyShape {
+        // TODO[mm] cut down on clone usage
+
+        let inner_radius : Var<f32> = spec.inner_radius.into();
+        let height       : Var<f32> = spec.height.clone().into();
+
+        let segment_width_rad: Var<f32>       = spec.width.clone().into();
+        let segment_radius   : Var<f32>       = inner_radius.clone().into();
+        let segment : CircleSegment<Var<f32>> = CircleSegment::new(segment_radius,segment_width_rad);
+
+        let tri_base             = segment.sagitta();
+        let tri_height: Var<f32> = &height + &tri_base;
+        let tri_width            = segment.chord_length();
+
+        // TODO[mm]] consider replace triangle with a `Plane().cut_angle`
+        // But avoid visual artifacts at the other end of the circle.
+        // let section = Plane().cut_angle(&spec.width);
+        // let section = section.rotate(180.0.deg().radians());
+        // let section = section.translate_y(tri_offset);
+
+        let triangle = Triangle(&tri_width, &tri_height);
+
+        let circle_radius: Var<Distance<Pixels>> = inner_radius.clone().into();
+        let circle_inner    = Circle(circle_radius);
+
+        let circle_offset_y: Var<Distance<Pixels>> = (&tri_base-&inner_radius).into();
+        let circle_inner    = circle_inner.translate_y(circle_offset_y);
+
+        let triangle_rounded = Difference(triangle,circle_inner);
+        let triangle_rounded = triangle_rounded.fill(Srgb::new(0.26, 0.69, 0.99));
+
+        let tri_offset: Var<Distance<Pixels>> = (-&tri_base).into();
+        let triangle_rounded = triangle_rounded.translate_y(tri_offset);
+
+        triangle_rounded.into()
+    }
+
+    /// Canvas node shape definition.
+    ensogl::define_shape_system! {
+        (height:f32,width:f32,inner_radius:f32) {
+            // TODO[mm] take spec or spec values as `Var<_>` parameters
+            let port_spec_val = SpecificationVar{
+                    height       : height.into(),
+                    width        : width.into(),
+                    inner_radius : inner_radius.into()
+            };
+            new_port_outwards(port_spec_val)
+        }
+    }
+}
+
+
+// =================
+// === Port Node ===
+// =================
+
+// TODO[mm] remove and use values derived from node instead
+const DEFAULT_WIDTH  : f32 = 25.0 * (PI / 180.0);
+const DEFAULT_RADIUS : f32 = 60.0;
+const DEFAULT_HEIGHT : f32 = 30.0;
+
+
+/// TODO consider unifying the ShapeViews.
+/// Shape view for Input Port.
+#[derive(Debug,Clone,Copy)]
+pub struct InputPortView {}
+impl component::ShapeViewDefinition for InputPortView {
+    type Shape = shape_in::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene, _shape_registry:&ShapeRegistry) -> Self {
+        shape.width.set(DEFAULT_WIDTH);
+        shape.inner_radius.set(DEFAULT_RADIUS);
+        shape.height.set(DEFAULT_HEIGHT);
+
+        shape.sprite.size().set(Vector2::new(200.0,200.0));
```

Comment:
do we need such a big ports?

Reply:
Only temprarily. I need to implement some re-centring of the shape and calculate this dynamically, depending on the actual size of the shape. That way we can minimise the sprite size.